polar-core/src/resource_block.rs

```diff
@@ -271,18 +290,56 @@ impl ResourceBlocks {
         resource: Term,
         declarations: Declarations,
         shorthand_rules: Vec<ShorthandRule>,
-    ) {
-        self.declarations.insert(resource.clone(), declarations);
-        self.shorthand_rules
-            .insert(resource.clone(), shorthand_rules);
+    ) -> Result<()> {
+        // Merge existing declarations if we are reopening a resource block, otherwise add new
+        if let Some(existing) = self.declarations.get_mut(&resource) {
+            for (key, new) in &declarations {
+                let existing_declaration = existing.get(key);
+                // allow
+                // 1. overwriting declarations of the **same** type
+                // 2. inserting declarations where none previously existed
+                //
+                // deny overwriting of declarations with mismatched types
+                match (new, existing_declaration) {
+                    // new or same-type declarations are accepted
+                    (Declaration::Relation(_), Some(Declaration::Relation(_)))
+                    | (Declaration::Relation(_), None)
+                    | (Declaration::Role, Some(Declaration::Role))
+                    | (Declaration::Role, None)
+                    | (Declaration::Permission, Some(Declaration::Permission))
+                    | (Declaration::Permission, None) => {
+                        existing.insert(key.clone(), new.clone());
+                    }
+                    // disallow all other combinations (mismatched declaration types)
+                    _ => {
+                        return Err(ValidationError::DuplicateResourceBlockDeclaration {
+                            resource,
+                            declaration: key.clone(),
+                            existing: existing_declaration.unwrap().clone(),
+                            new: new.clone(),
+                        })
```

Comment:
In the spirit of surfacing multiple errors at once, it would be nice to not bomb out when encountering the first error here. Instead, we should collect errors up and then _always_ return a `Vec<ValidationError>` from `ResourceBlocks::add`

Reply:
great shout. updated the function signature here to return `Vec<ValidationError>`